pennylane/devices/default_qutrit.py

```diff
@@ -0,0 +1,258 @@
+        consecutive_wires = range(self.num_wires)
+        wire_map = zip(wires, consecutive_wires)
+        return dict(wire_map)
+
+    def apply(self, operations, rotations=None, **kwargs):  # pylint: disable=arguments-differ
+        rotations = rotations or []
+
+        # apply the circuit operations
+        for i, operation in enumerate(operations):  # pylint: disable=unused-variable
+            self._state = self._apply_operation(self._state, operation)
+
+        # store the pre-rotated state
+        self._pre_rotated_state = self._state
+
+        # apply the circuit rotations
+        for operation in rotations:
+            self._state = self._apply_operation(self._state, operation)
+
+    def _apply_operation(self, state, operation):
+        """Applies operations to the input state.
+
+        Args:
+            state (array[complex]): input state
+            operation (~.Operation): operation to apply on the device
+
+        Returns:
+            array[complex]: output state
+        """
+        if operation.base_name == "Identity":
+            return state
+        wires = operation.wires
+
+        if operation.base_name in self._apply_ops:
+            axes = self.wires.indices(wires)
+            return self._apply_ops[operation.base_name](state, axes, inverse=operation.inverse)
+
+        matrix = self._asarray(self._get_unitary_matrix(operation), dtype=self.C_DTYPE)
+
+        return self._apply_unitary(state, matrix, wires)
+
+    def _get_unitary_matrix(self, unitary):  # pylint: disable=no-self-use
+        """Return the matrix representing a unitary operation.
+
+        Args:
+            unitary (~.Operation): a PennyLane unitary operation
+
+        Returns:
+            array[complex]: Returns a 2D matrix representation of
+            the unitary in the computational basis.
+        """
+        return unitary.matrix()
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qutrit",
+            supports_reversible_diff=True,
+            supports_inverse_operations=True,
+            supports_analytic_computation=True,
+            returns_state=True,
+        )
+        return capabilities
+
+    def _create_basis_state(self, index):
+        """Return a computational basis state over all wires.
+
+        Args:
+            index (int): integer representing the computational basis state
+
+        Returns:
+            array[complex]: complex array of shape ``[3]*self.num_wires``
+            representing the statevector of the basis state
+        """
+        state = np.zeros(3**self.num_wires, dtype=np.complex128)
+        state[index] = 1
+        state = self._asarray(state, dtype=self.C_DTYPE)
+        return self._reshape(state, [3] * self.num_wires)
+
+    @property
+    def state(self):
+        return self._flatten(self._pre_rotated_state)
+
+    def density_matrix(self, wires):
```

Comment:
Would it make more sense to put this in the definition of `QutritDevice` to match how it is done for `QubitDevice`? Right now, the base class will still raise a `QuantumFunctionError`, but it looks now like this will be supported just fine.

Reply:
Ah I didn't realize this got changed. `density_matrix` used to be in `DefaultQubit`, but I guess it got moved to `QubitDevice` when `qml.math` got support for `reduced_dm`. I can't use `qml.math.reduced_dm` as it is hard-coded to support qubits. So I will keep this implementation for now and move it to `QutritDevice`.

Reply:
Update: I have reverted this change and put the working implementation of `density_matrix` back into `DefaultQutrit`. I didn't look carefully enough when I was checking if the method could be moved to `QutritDevice`. `DefaultQutrit.density_matrix` has a reference to `self._prerotated_state`, which is a class variable that is NOT inherited from `QutritDevice`, so it needs to stay in `DefaultQutrit`, at least until `qml.math` is updated, or until I change the variables of `QutritDevice`, which I'm hesitant to do as I think it's best to stick to the same format as `QubitDevice`.

Reply:
Yes I think this was moved in part due to the quantum information module.
Options are to have a separate method here or modify the separated function to allow for qutrits 🤔? I think for the moment best to stick with the former, as you are doing right now. At the moment qutrits are still quite experimental, until that changes its fine to have this separate method here.